src/sentry_value.c

```diff
@@ -1036,8 +1038,38 @@ sentry_value_new_breadcrumb(const char *type, const char *message)
     return rv;
 }
 
-void
-sentry_event_value_add_stacktrace(sentry_value_t event, void **ips, size_t len)
+sentry_value_t
+sentry_value_new_exception(const char *type, const char *value)
+{
+    sentry_value_t exc = sentry_value_new_object();
+    sentry_value_set_by_key(exc, "type", sentry_value_new_string(type));
+    sentry_value_set_by_key(exc, "value", sentry_value_new_string(value));
+    return exc;
+}
+
+sentry_value_t
+sentry_value_new_thread(uint64_t id, const char *name)
+{
+    sentry_value_t thread = sentry_value_new_object();
+
+    // TODO: would be nice to be able to actually use a `u64` as value.
+    char buf[100];
+    size_t written
+        = (size_t)snprintf(buf, sizeof(buf), "%llu", (unsigned long long)id);
+    if (written < sizeof(buf)) {
+        buf[written] = '\0';
```

Comment:
This has multiple issues:
- The return value of `snprintf` is not how many characters were written, but how many would have been written. The name is misleading.
- `snprintf` writes the NUL byte for you in all cases according to the standard.
- if `written == sizeof(buf)`, this overflow and writes past the buffer.

Please audit remaining code for similar issues.

Reply:
Its the same code as in https://github.com/getsentry/sentry-native/pull/508 that I landed recently.
`written` will never be `== sizeof(buf)` per the if. And yes, this is written very defensively, as I’m not sure which parts of printf we can actually trust.

Reply:
My mistake, you're right 👍